include/gabia/secure_socket.hpp

```diff
@@ -0,0 +1,394 @@
+
+        void operator()(beast::error_code error, size_t n) {
+            CORO_REENTER(*this) {
+                static_cast<void>(n);
+
+                state_ptr->socket.encrypt_pdu(state_ptr->sequence,
+                                              state_ptr->pdu);
+                CORO_YIELD asio::async_write(state_ptr->socket.next_layer(),
+                                             asio::buffer(state_ptr->pdu),
+                                             *this);
+                auto n = state_ptr->pdu.size() - sizeof(uint16_t) -
+                         crypto::aead_auth_tag_size;
+                state_ptr.invoke(error, n);
+            }
+        }
+
+        friend void* asio_handler_allocate(std::size_t size,
+                                           async_write_some_op* op) {
+            using boost::asio::asio_handler_allocate;
+            return asio_handler_allocate(
+                size, std::addressof(op->state_ptr.handler()));
+        }
+
+        friend void asio_handler_deallocate(void* p, std::size_t size,
+                                            async_write_some_op* op) {
+            using boost::asio::asio_handler_deallocate;
+            asio_handler_deallocate(p, size,
+                                    std::addressof(op->state_ptr.handler()));
+        }
+
+        friend bool asio_handler_is_continuation(async_write_some_op* op) {
+            using boost::asio::asio_handler_is_continuation;
+            return asio_handler_is_continuation(
+                std::addressof(op->state_ptr.handler()));
+        }
+
+        template <class Function>
+        friend void asio_handler_invoke(Function&& f, async_write_some_op* op) {
+            using boost::asio::asio_handler_invoke;
+            asio_handler_invoke(f, std::addressof(op->state_ptr.handler()));
+        }
+
+        struct state {
+            state(Handler& handler, secure_socket& socket,
+                  const ConstBufferSequence& sequence)
+                : socket{socket},
+                  sequence{sequence},
+                  pdu{beast::handler_alloc<char, Handler>{handler}} {}
+
+            secure_socket& socket;
+            ConstBufferSequence sequence;
+            using alloc_t = beast::handler_alloc<gsl::byte, Handler>;
+            std::vector<gsl::byte, alloc_t> pdu;
+        };
+
+        beast::handler_ptr<state, Handler> state_ptr;
+    };
+
+    template <typename MutableBufferSequence, typename Handler>
+    struct async_read_some_op : public asio::coroutine {
+        template <typename DeducedHandler>
+        async_read_some_op(secure_socket& socket,
+                           const MutableBufferSequence& sequence,
+                           DeducedHandler&& handler)
+            : socket{socket},
+              sequence{sequence},
+              handler{std::forward<DeducedHandler>(handler)} {}
+        secure_socket& socket;
+        MutableBufferSequence sequence;
+        Handler handler;
```

Comment:
I *think* you need
```
async_read_some_op(async_read_some_op&&) = default;
async_read_some_op(async_read_some_op const&) = default;
```

Otherwise it will always copy and never move. In general, you should be explicit with special members.